src/calendar/Calendar.ts

```diff
@@ -0,0 +1,391 @@
+	protected onPropertiesChanged(evt: PropertiesChangeEvent<this, CalendarProperties>) {
+		const { customDateCell = CalendarCell } = this.properties;
+
+		// update custom option registry
+		if ( !this._registry || includes(evt.changedPropertyKeys, 'customDateCell')) {
+			const registry = this._createRegistry(customDateCell);
+
+			if (this._registry) {
+				this.registries.replace(this._registry, registry);
+			}
+			else {
+				this.registries.add(registry);
+			}
+			this._registry = registry;
+		}
+	}
+
+	private _createRegistry(customDateCell: any) {
+		const registry = new WidgetRegistry();
+		registry.define('date-cell', customDateCell);
+
+		return registry;
+	}
+
+	private _getMonthLength(month: number, year: number) {
+		const lastDate = new Date(year, month + 1, 0);
+		return lastDate.getDate();
+	}
+
+	private _getMonthYear() {
+		const {
+			month,
+			selectedDate = this._defaultDate,
+			year
+		} = this.properties;
+		return { month: typeof month === 'number' ? month : selectedDate.getMonth(), year: year ? year : selectedDate.getFullYear() };
+	}
+
+	private _goToDate(day: number) {
+		const {
+			month,
+			year
+		} = this._getMonthYear();
+		const currentMonthLength = this._getMonthLength(month, year);
+		const previousMonthLength = this._getMonthLength(month - 1, year);
+
+		if (day < 1) {
+			this._onMonthDecrement();
+			day += previousMonthLength;
+		}
+		else if (day > currentMonthLength) {
+			this._onMonthIncrement();
+			day -= currentMonthLength;
+		}
+
+		this._focusedDay = day;
+		this._callDateFocus = true;
+		this.invalidate();
+	}
+
+	private _onDateClick(date: number, disabled: boolean) {
+		const { onDateSelect } = this.properties;
+		let {
+			month,
+			year
+		} = this._getMonthYear();
+
+		if (disabled && date < 15) {
+			({ month, year } = this._onMonthIncrement());
+			this._callDateFocus = true;
+		}
+		else if (disabled && date >= 15) {
+			({ month, year } = this._onMonthDecrement());
+			this._callDateFocus = true;
+		}
+		this._focusedDay = date;
+		onDateSelect && onDateSelect(new Date(year, month, date));
+	}
+
+	private _onDateFocusCalled() {
+		this._callDateFocus = false;
+	}
+
+	private _onDateKeyDown(event: KeyboardEvent) {
+		const { month, year } = this._getMonthYear();
+		switch (event.which) {
+			case Keys.Up:
+				this._goToDate(this._focusedDay - 7);
+				break;
+			case Keys.Down:
+				this._goToDate(this._focusedDay + 7);
+				break;
+			case Keys.Left:
+				this._goToDate(this._focusedDay - 1);
+				break;
+			case Keys.Right:
+				this._goToDate(this._focusedDay + 1);
+				break;
+			case Keys.PageUp:
+				this._goToDate(1);
+				break;
+			case Keys.PageDown:
+				const monthLengh = this._getMonthLength(month, year);
+				this._goToDate(monthLengh);
+				break;
+			case Keys.Enter:
+			case Keys.Space:
+				const { onDateSelect } = this.properties;
+				onDateSelect && onDateSelect(new Date(year, month, this._focusedDay));
+		}
+	}
+
+	private _onMonthDecrement() {
```

Comment:
I'll be perpetually stuck in a loop of splitting them out and merging them again 😂 .
I'm really fine either way, they seem on the edge of annoying repetitiveness vs. annoying conditionals